main/src/library/Array.flix

```diff
@@ -57,10 +57,1390 @@ namespace Array {
+    /// searching from left to right.
+    ///
+    pub def indexOfLeft(x: a, a: Array[a]): Option[Int32] & Pure =
+        let i = indexOfLeftHelper(x, a, length(a), 0) as & Pure;
+        if (i < 0) None else Some(i)
+
+    ///
+    /// Helper function for `indexOfLeft`.
+    ///
+    def indexOfLeftHelper(x: a, a: Array[a], len:Int32, i: Int32): Int32 & Impure =
+        if (i >= len)
+            -1
+        else
+            if (a[i] == x) i else indexOfLeftHelper(x, a, len, i+1)
+
+    ///
+    /// Optionally returns the position of the first occurence of `x` in `a`
+    /// searching from right to left.
+    ///
+    pub def indexOfRight(x: a, a: Array[a]): Option[Int32] & Pure =
+        let start = length(a) - 1;
+        let i = indexOfRightHelper(x, a, start) as & Pure;
+        if (i < 0) None else Some(i)
+
+    ///
+    /// Helper function for `indexOfRight`.
+    ///
+    def indexOfRightHelper(x: a, a: Array[a], i: Int32): Int32 & Impure =
+        if (i < 0)
+            -1
+        else
+            if (a[i] == x) i else indexOfRightHelper(x, a, i-1)
+
+    ///
+    /// Return the positions of the all the occurences of `x` in `a`.
+    ///
+    pub def indices(x: a, a: Array[a]): Array[Int32] & Impure =
+        findIndices(y -> x == y, a)
+
+    ///
+    /// Alias for `findLeft`.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def find(f: a -> Bool, xs: Array[a]): Option[a] = findLeft(f, xs)
+
+    ///
+    /// Optionally returns the first element of `a` that satisfies the predicate `f` when searching from left to right.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def findLeft(f: a -> Bool, a: Array[a]): Option[a] & Pure =
+        match findIndexOfLeft(x -> f(x), a) {
+            case None => None
+            case Some(i) => Some(a[i] as & Pure)
+        }
+
+    ///
+    /// Optionally returns the first element of `xs` that satisfies the predicate `f` when searching from right to left.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def findRight(f: a -> Bool, a: Array[a]): Option[a] =
+        match findIndexOfRight(x -> f(x), a) {
+            case None => None
+            case Some(i) => Some(a[i] as & Pure)
+        }
+
+    ///
+    /// Returns an array of all integers between `b` (inclusive) and `e` (exclusive).
+    ///
+    /// Returns `[]` if `b >= e`.
+    ///
+    pub def range(b: Int32, e: Int32): Array[Int32] & Impure =
+        rangeHelper(b,e)
+
+    /// 
+    /// Helper for `range`.
+    ///
+    def rangeHelper(b: Int32, e: Int32): Array[Int32] & Impure =
+        if (b >= e)
+            []
+        else {
+            let f = x -> x + b;
+            init(f, e-b)
+        }
+
+    ///
+    /// Returns an array with the element `x` repeated `n` times.
+    ///
+    /// Returns `[]` if `n <= 0`.
+    ///
+    pub def repeat(x: a, n: Int32): Array[a] & Impure =
+        if (n > 0)
+            new(x, n)
+        else
+            []
```

Comment:
I most other functions you have the "base" case first and then the recursive case afterwards.